src/daft-scan/src/scan_task_iters.rs

```diff
@@ -0,0 +1,105 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+
+use crate::{ScanTask, ScanTaskRef};
+
+type BoxScanTaskIter = Box<dyn Iterator<Item = DaftResult<ScanTaskRef>>>;
+
+/// Coalesces ScanTasks by their [`ScanTask::size_bytes()`]
+///
+/// NOTE: `min_size_bytes` and `max_size_bytes` are only parameters for the algorithm used for merging ScanTasks,
+/// and do not provide any guarantees about the sizes of ScanTasks yielded by the resultant iterator.
+/// This function may still yield ScanTasks with smaller sizes than `min_size_bytes`, or larger sizes
+/// than `max_size_bytes` if **existing non-merged ScanTasks** with those sizes exist in the iterator.
+///
+/// # Arguments:
+///
+/// * `scan_tasks`: A Boxed Iterator of ScanTaskRefs to perform merging on
+/// * `min_size_bytes`: Minimum size in bytes of a ScanTask, after which no more merging will be performed
+/// * `max_size_bytes`: Maximum size in bytes of a ScanTask, capping the maximum size of a merged ScanTask
+pub fn merge_by_sizes(
+    scan_tasks: BoxScanTaskIter,
+    min_size_bytes: usize,
+    max_size_bytes: usize,
+) -> BoxScanTaskIter {
+    Box::new(MergeByFileSize {
+        iter: scan_tasks,
+        min_size_bytes,
+        max_size_bytes,
+        accumulator: None,
+    })
+}
+
+struct MergeByFileSize {
+    iter: BoxScanTaskIter,
+    min_size_bytes: usize,
+    max_size_bytes: usize,
+
+    // Current element being accumulated on
+    accumulator: Option<ScanTaskRef>,
+}
+
+impl Iterator for MergeByFileSize {
+    type Item = DaftResult<ScanTaskRef>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Grabs the accumulator, leaving a `None` in its place
+        let accumulator = self.accumulator.take();
+
+        match (self.iter.next(), accumulator) {
+            // When no accumulator exists, trivially place the ScanTask into the accumulator
+            (Some(Ok(child_item)), None) => {
+                self.accumulator = Some(child_item);
+                self.next()
```

Comment:
I would put this match in a loop instead of recursion. Since the stack depth will be the number of files you are fusing and will make debugging difficult 
```
loop {
        match (self.iter.next(), accumulator) {
            // When no accumulator exists, trivially place the ScanTask into the accumulator
            (Some(Ok(child_item)), None) => {
                self.accumulator = Some(child_item);
                continue;
            }
```

Reply:
Makes sense